rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+
+For this initial implementation, `stdio` will be used between the AI tool and
+`tsh`'s local proxy, and between the App service and the MCP server getting
+launched on that host.
+
+`tsh` will use a local proxy and starts a TLS routing tunnel with the Proxy,
+using a new ALPN value `teleport-mcp`. Technically, the initial implementation
+can reuse ALPN `teleport-tcp` as TCP and MCP apps share the same app handler
+entrypoint. However, a new ALPN is introduced in case this entry point changes
+for MCP servers in the future.
+
+The duplex channel of the tunnelled TLS-routing connection to the Proxy will
+forward stdin and stdout respectively.
+
+#### Implementation - auth
+
+Since MCP servers are registered as apps, existing role option `app_labels` is
+used to control which MCP servers a role can access based on the labels of the
+MCP server apps.
+
+A new role option `mcp_tools` is introduced to apply fine-grained control on
+which tools from the MCP server should be allowed. Similar to
+`kubernetes_resources.name`, entries in `mcp_tools` also support glob and regex
+representations to make whitelisting or blacklisting easier. If `mcp_tools` is not
+specified for a role, all tools are allowed by default. This new option
+`mcp_tools` will be enforced per JSON-RPC `tool/call` call, which will be
+detailed in the next section.
+
+The procedure to access an MCP server from `tsh` is the same as any other app --
+a user has to obtain a user certificate with `route_to_app` for the TLS routing
+handshake, and the backend validates the identity then routes to the appropriate
+app service based on this certificate.
+
+#### Implementation - MCP server handler with JSON-RPC
+
+MCP server handler will be placed in App service's connection handler. The MCP
+server handler will receive a raw connection once the connection is authorized
+and confirmed to be an MCP app. It is expected [the MCP protocol in JSON-RPC
+format](https://modelcontextprotocol.io/specification/2025-03-26) will be
+transferred within this connection.
+
+Since using `stdio`, a `exec.Command` defined by the app definition is executed
+to start an MCP server, for each authorized incoming client connection. Then
+stdin and stdout are proxied between the client connection and the launched MCP
+server. Stderr from the launched MCP server will be logged at TRACE level in
+Teleport logs.
+
+Incoming bytes from the client do get the following treatments before they are
+forwarded to the launched MCP server:
+- Parsed into MCP/JSON-RPC messages. 
+- If the message is a `tools/call` request, use the access-checker to confirm
+  whether the identity has access to the requested tool.
+  - If the `tools/call` is access denied, make a text result detailing this
+    error and send it back to the client connection. Do not forward the request to
+    the launched MCP server in this case.
+- Record audit events for eligible messages.
+
+The launched MCP server should be stopped once the connection ends.
```

Comment:
Another thing to consider is monitoring MCP server behaviour. Capturing all of the session info might be too much (unless we want this?), but perhaps a summary wouldn't be too bad.